authority/provisioner/azure.go

```diff
@@ -444,3 +476,33 @@ func (p *Azure) assertConfig() {
 		p.config = newAzureConfig(p.TenantID)
 	}
 }
+
+// getAzureEnvironment returns the Azure environment for the current instance
+func (p *Azure) getAzureEnvironment() (string, error) {
+	req, err := http.NewRequest("GET", p.config.instanceComputeURL, http.NoBody)
+	if err != nil {
+		return "", errors.Wrap(err, "error creating request")
+	}
+	req.Header.Add("Metadata", "True")
+
+	query := req.URL.Query()
+	query.Add("format", "text")
+	query.Add("api-version", "2021-02-01")
+	req.URL.RawQuery = query.Encode()
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", errors.Wrap(err, "error getting azure instance environment, are you in a Azure VM?")
+	}
+	defer resp.Body.Close()
+
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", errors.Wrap(err, "error reading azure environment response")
+	}
+	if resp.StatusCode >= 400 {
+		return "", errors.Errorf("error getting azure environment: status=%d, response=%s", resp.StatusCode, b)
+	}
+
+	return string(b), nil
+}
```

Comment:
Although `GetIdentityToken()` is only used in the CLI, we should only do this once and have a cached version of the environment or the resource. It might not be a problem for `step`, but not having a cached version might cause unnecessary requests if this is used on a long-running app.

Reply:
That makes sense, probably caching it within the `Azure` struct unless you have a better suggestion.

Reply:
Yes caching it there.

Reply:
Addressed in [09cbe8b](https://github.com/smallstep/certificates/pull/1309/commits/09cbe8ba653c26b2bcb5ee6ec09a2144817423d0)